### PR TITLE
Implement determineAvailableVehicles

### PR DESCRIPTION
**Pull request by duncanspumpkin**

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
Oh lol this is broken forgot to push the right code

File: src/OpenLoco/src/Vehicles/VehicleManager.cpp [hunk 1]
```diff
@@ -56,12 +57,63 @@ namespace OpenLoco::VehicleManager
...
     }
 
+    // 0x004C39D4
+    static void determineAvailableVehicleTypes(Company& company)
+    {
+        uint16_t availableTypes = 0;
+
+        for (uint32_t i = 0; i < ObjectManager::getMaxObjects(ObjectType::vehicle); ++i)
+        {
+            if (!(company.unlockedVehicles[i & 0x1F] & 1ULL << (i / 32)))
```
Comment (ZehMatt) [inline src/OpenLoco/src/Vehicles/VehicleManager.cpp:67]: It's probably better to create a new variable for the mask, the member is also 32 bit and not 64.

File: src/OpenLoco/src/Vehicles/VehicleManager.cpp [hunk 2]
```diff
@@ -56,12 +57,63 @@ namespace OpenLoco::VehicleManager
...
+                continue;
+            }
+
+            auto* vehObj = ObjectManager::get<VehicleObject>(i);
+            if (vehObj == nullptr)
+            {
+                continue;
+            }
+
+            availableTypes |= 1ULL << enumValue(vehObj->type);
```
Comment (ZehMatt) [inline src/OpenLoco/src/Vehicles/VehicleManager.cpp:78]:
```suggestion
            availableTypes |= 1U << enumValue(vehObj->type);
```
availableTypes here is only 16 bits

File: src/OpenLoco/src/Vehicles/VehicleManager.cpp [hunk 3]
```diff
@@ -56,12 +57,63 @@ namespace OpenLoco::VehicleManager
...
+                continue;
+            }
+
+            const auto forbiddenVehicles = CompanyManager::isPlayerCompany(company.id()) ? getGameState().forbiddenVehiclesPlayers : getGameState().forbiddenVehiclesCompetitors;
+            if (forbiddenVehicles & (1ULL << enumValue(vehObj->type)))
+            {
+                continue;
+            }
+
+            company.unlockedVehicles[i & 0x1F] |= 1ULL << (i / 32);
```
Comment (ZehMatt) [inline src/OpenLoco/src/Vehicles/VehicleManager.cpp:113]:
```suggestion
            company.unlockedVehicles[i & 0x1F] |= 1U << (i / 32);
```

**Review by LeftofZen — COMMENTED** [review]
Playtested, didn't see any difference to current master. I think there is one comment to fix from ZehMatt and then we can merge this

File: src/OpenLoco/src/Vehicles/VehicleManager.cpp [hunk 4]
```diff
@@ -56,12 +57,66 @@ namespace OpenLoco::VehicleManager
         }
     }
 
+    // 0x004C39D4
+    static uint16_t determineAvailableVehicleTypes(Company& company)
```
Comment (ZehMatt) [inline src/OpenLoco/src/Vehicles/VehicleManager.cpp:61]:
```suggestion
    static uint16_t determineAvailableVehicleTypes(const Company& company)
```

File: src/OpenLoco/src/Vehicles/VehicleManager.cpp [hunk 5]
```diff
@@ -56,12 +57,66 @@ namespace OpenLoco::VehicleManager
...
+            {
+                continue;
+            }
+
+            availableTypes |= 1U << enumValue(vehObj->type);
+        }
+        return availableTypes;
+    }
+
+    static BitSet<224> determineUnlockedVehicles(Company& company)
```
Comment (ZehMatt) [inline src/OpenLoco/src/Vehicles/VehicleManager.cpp:83]:
```suggestion
    static BitSet<224> determineUnlockedVehicles(const Company& company)
```

**Review by ZehMatt — APPROVED** [review]

**Pull request merged**